hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/SevenToEightUpgradeHandler.java

```diff
@@ -178,23 +179,68 @@ static void upgradePartitionFields(HoodieWriteConfig config, HoodieTableConfig t
   }
 
   static void upgradeMergeMode(HoodieTableConfig tableConfig, Map<ConfigProperty, String> tablePropsToAdd) {
-    if (tableConfig.getPayloadClass() != null
-        && tableConfig.getPayloadClass().equals(OverwriteWithLatestAvroPayload.class.getName())) {
-      if (HoodieTableType.COPY_ON_WRITE == tableConfig.getTableType()) {
+    String payloadClass = tableConfig.getPayloadClass();
+    String preCombineField = tableConfig.getPreCombineField();
+    if (isCustomPayloadClass(payloadClass)) {
+      // This contains a special case: HoodieMetadataPayload.
+      tablePropsToAdd.put(
+          HoodieTableConfig.PAYLOAD_CLASS_NAME,
+          payloadClass);
+      tablePropsToAdd.put(
+          HoodieTableConfig.RECORD_MERGE_MODE,
+          RecordMergeMode.CUSTOM.name());
+      tablePropsToAdd.put(
+          HoodieTableConfig.RECORD_MERGE_STRATEGY_ID,
+          HoodieRecordMerger.PAYLOAD_BASED_MERGE_STRATEGY_UUID);
+    } else if (StringUtils.nonEmpty(preCombineField)) {
+      // This contains a special case: OverwriteWithLatestPayload with preCombine field.
+      tablePropsToAdd.put(
+          HoodieTableConfig.PAYLOAD_CLASS_NAME,
+          DefaultHoodieRecordPayload.class.getName());
```

Comment:
won't we run into Table config validation issue here? i.e for payload class, before upgrade it could be OverwriteWithLatest and post upgrade its DefaultHoodieRecordPayload. 

I am not suggesting to revert this. But just curious if there are chances of exception.

Reply:
oh, is it possible to leave the payload class as is, but just set the merge mode to EventTime based merge mode? 
if not, we might break existing pipelines. 
post upgrade, user's write config could have OverwriteWithLatest, but table config might say as DefaultHoodieRecordPayload and table config validation might fail.

Reply:
@yihua : is this a valid state for a given table config? and if yes, which ones takes precedence. 
i.e. payload class set to OverwriteWithLatest
and merge mode set to EventTimebased merge mode

Reply:
For table version 8, the merge mode, payload class and custom merger configs (strategy ID and implementation class) are always consistent.  So the upgrade process should maintain that.

Reply:
for COW, we will try to keep Commit time if its OverwriteWithLatest. (both precomb set or not)
MOR: follow whats in the patch. 

lets relax and do warn log for payload mismatch during tbale config validation. 
also, during inference if there is payload mis-alignment